pkg/cache/cache.go

```diff
@@ -147,31 +134,36 @@ type ViewOptions struct {
 	DefaultTransform toolscache.TransformFunc
 
 	// ByObject restricts the cache's ListWatch to the desired fields per GVK at the specified object.
-	ByObject ViewByObject
+	ByObject map[client.Object]ByObject
+
+	// UnsafeDisableDeepCopy indicates not to deep copy objects during get or
+	// list objects for EVERY object.
+	// Be very careful with this, when enabled you must DeepCopy any object before mutating it,
+	// otherwise you will mutate the object in the cache.
+	UnsafeDisableDeepCopy *bool
 }
 
-// ViewByObject offers more fine-grained control over the cache's ListWatch by object.
-type ViewByObject struct {
-	// Selectors restricts the cache's ListWatch to the desired
-	// fields per GVK at the specified object, the map's value must implement
-	// Selectors [1] using for example a Set [2]
-	// [1] https://pkg.go.dev/k8s.io/apimachinery/pkg/fields#Selectors
-	// [2] https://pkg.go.dev/k8s.io/apimachinery/pkg/fields#Set
-	Selectors SelectorsByObject
+// ByObject offers more fine-grained control over the cache's ListWatch by object.
+type ByObject struct {
+	// Label represents a label selector for the object.
+	Label labels.Selector
+
+	// Field represents a field selector for the object.
+	Field fields.Selector
 
 	// Transform is a map from objects to transformer functions which
 	// get applied when objects of the transformation are about to be committed
 	// to cache.
 	//
 	// This function is called both for new objects to enter the cache,
 	// and for updated objects.
-	Transform TransformByObject
+	Transform toolscache.TransformFunc
 
 	// UnsafeDisableDeepCopy indicates not to deep copy objects during get or
 	// list objects per GVK at the specified object.
 	// Be very careful with this, when enabled you must DeepCopy any object before mutating it,
 	// otherwise you will mutate the object in the cache.
-	UnsafeDisableDeepCopy DisableDeepCopyByObject
+	UnsafeDisableDeepCopy bool
```

Comment:
@vincepri just curious, is it intentional to not overwrite `UnsafeDisableDeepCopy` by the top level `opts. UnsafeDisableDeepCopy` if it's not specified per object? (I understand its safe to default it to false, but just asking if it would be confusing to the user). Also is there a reason to have it as a pointer in top level and not in byObject?

Reply:
The thinking behind this was to retain some of the features we have today, and make it a bit safer as you said. 

- If a user wants to disable deep copy globally, we should be able to override it for specific objects.
- If users want to keep it to false globally (default), we should be able to override it to true for specific objects.

The reason for the top level pointer is for merging options together, there is a need to know if the option was left unset, or was false, which is the default of a boolean field. The option inheritance logic is hard to follow (spent a good deal of time on it today 😄), but it basically requires the pointer to know if we should inherit or not.

Happy to chat about it more if we feel like this should be made safer or easier to use. I would expect though that most users shouldn't disable deep copy unless they absolutely know the repercussions of doing so, which is an advanced use case

Reply:
Ah, I see, got it! Thanks for explaining @vincepri :)